boards/risc-v/esp32c3-legacy/common/src/esp32c3_board_twai.c

```diff
@@ -24,78 +27,615 @@
+ * memory must be writable because, under certain circumstances, the driver
+ * may modify frequency or X plate resistance values.
+ */
+
+static struct esp32c3_twai_config_s
+    g_esp32c3_twai_config[] =
+{
+#ifdef CONFIG_ESP32C3_TWAI0
+    {
+      .config =
+        {
+          .bittiming_const = &esp32c3_twai_bittiming_const,
+          .port     = 0,
+          .periph   = ESP32C3_PERIPH_TWAI,
+          .irq      = ESP32C3_IRQ_TWAI,
+          .cpuint   = -ENOMEM,
+          .bitrate  = CONFIG_CAN_SJA1000_BITRATE,
+          .samplep  = CONFIG_CAN_SJA1000_SAMPLEP,
+          .sjw      = CONFIG_CAN_SJA1000_SJW,
+#ifdef CONFIG_CAN_LOOPBACK
+          .loopback = true,
+#else
+          .loopback = false,
+#endif /* #ifdef CONFIG_CAN_LOOPBACK */
+          .triple_sample = CONFIG_CAN_SJA1000_SAM,
+          .attach        = twai_attach,
+          .detach        = twai_detach,
+        },
+    },
+#endif /* #ifdef CONFIG_ESP32C3_TWAI0 */
+#ifdef CONFIG_ESP32C3_TWAI1
+#  error "TODO - TWAI1 not yet defined"
+#endif /* #ifdef CONFIG_ESP32C3_TWAI1 */
+};
+
+static struct sja1000_dev_s g_twai_priv[] =
+{
+#ifdef CONFIG_ESP32C3_TWAI0
+    {
+      .getreg = twai_getreg,
+      .putreg = twai_putreg,
+      .modifyreg32 = twai_modifyreg32,
+      .config = &g_esp32c3_twai_config[0].config,
+      .base = DR_REG_TWAI_BASE,
+    },
+#endif /* #ifdef CONFIG_ESP32C3_TWAI0 */
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp32c3_twai_interrupt
  *
  * Description:
- *   Configure the TWAI driver.
+ *   This is the TWAI Interrupt handler.
  *
  * Input Parameters:
- *   None.
+ *   irq - Interrupt number
+ *   context - interrupt context
+ *   arg - pointer to sja1000_dev_s (lower CAN driver) context.
  *
  * Returned Value:
- *   Zero (OK) is returned on success; A negated errno value is returned
- *   to indicate the nature of any failure.
+ *   OK on success or negated error value.
  *
  ****************************************************************************/
 
-int board_twai_setup(void)
+int esp32c3_twai_interrupt(int irq, void *context, void *arg)
 {
-#ifdef CONFIG_ESP32C3_TWAI0
-  struct can_dev_s *twai;
+  struct sja1000_dev_s *priv = (struct sja1000_dev_s *)arg;
+  struct esp32c3_twai_config_s *esp32c3_twai_config
+      = &g_esp32c3_twai_config[priv->config->port];
+
+  cantrace("Entered for CAN port %" PRId8 ".\n", priv->config->port);
+  DEBUGASSERT(priv != NULL);
+  DEBUGASSERT(esp32c3_twai_config != NULL);
+  DEBUGASSERT(esp32c3_twai_config->handler != NULL);
+  DEBUGASSERT(priv->config != NULL);
+  DEBUGASSERT(esp32c3_twai_config->arg);
+
+  esp32c3_twai_config->handler(priv->config, esp32c3_twai_config->arg);
+
+  return OK;
+}
+
```

Comment:
I am not sure why there is required so much indirection. I would expect that we can directly register common `sja1000_interrupt` there. It should receive  `struct sja1000_dev_s` as argument and this function should be  registered directly as ISR If you need some specific/extended config for some board then it should be defined to contain common one and can be extended in board specific handler wrapper by `container_of`.

On the other hand I agree with attach and detach and ability to setup there some integration specific `sja1000_interrupt` wrapper. But I would use same arguments for common SJA1000 interrupt ISR as for the wrappers.